Crash when cache evict

this is my log
```
[5796] 2016/12/26 09:43:03.939862 [INF] Starting nats-streaming-server[Nerve-Hub] version 0.3.6
[5796] 2016/12/26 09:43:04.020973 [INF] Starting nats-server version 0.9.6
[5796] 2016/12/26 09:43:04.021950 [INF] Starting http monitor on 127.0.0.1:8222
[5796] 2016/12/26 09:43:04.025860 [INF] Listening for client connections on 0.0.0.0:4222
[5796] 2016/12/26 09:43:04.026837 [INF] Server is ready
[5796] 2016/12/26 09:43:04.303394 [INF] STAN: Message store is FILE
[5796] 2016/12/26 09:43:04.303394 [INF] STAN: --------- Store Limits ---------
[5796] 2016/12/26 09:43:04.307302 [INF] STAN: Channels:                 1000
[5796] 2016/12/26 09:43:04.308280 [INF] STAN: -------- channels limits -------
[5796] 2016/12/26 09:43:04.310234 [INF] STAN:   Subscriptions:          1000 *
[5796] 2016/12/26 09:43:04.311211 [INF] STAN:   Messages     :      10000000
[5796] 2016/12/26 09:43:04.313166 [INF] STAN:   Bytes        :     976.56 MB *
[5796] 2016/12/26 09:43:04.314143 [INF] STAN:   Age          :     unlimited *
[5796] 2016/12/26 09:43:04.318052 [INF] STAN: --------------------------------
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x1 addr=0x10 pc=0x517903]

goroutine 10 [running]:
panic(0x85b900, 0xc04200a090)
        /home/travis/.gimme/versions/go1.7.4.linux.amd64/src/runtime/panic.go:500 +0x1af
github.com/nats-io/nats-streaming-server/stores.(*FileMsgStore).evictFromCache(0xc042038340, 0x1493fbe46c0257d4)
        /home/travis/gopath/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:2174 +0xb3
github.com/nats-io/nats-streaming-server/stores.(*FileMsgStore).backgroundTasks(0xc042038340)
        /home/travis/gopath/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:1987 +0x3b1
created by github.com/nats-io/nats-streaming-server/stores.(*FileStore).newFileMsgStore
        /home/travis/gopath/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:1252 +0xca5
```

I try to reproduce this crash and got one more log, there is one more line `Slow Consumer Detected` when I push around 16k msg nats server crash reliably with bellow log
```
7048] 2016/12/27 10:09:41.477593 [INF] Starting nats-streaming-server[Nerve-Hub] version 0.3.6
[7048] 2016/12/27 10:09:41.573362 [INF] Starting nats-server version 0.9.6
[7048] 2016/12/27 10:09:41.573362 [INF] Starting http monitor on 127.0.0.1:8222
[7048] 2016/12/27 10:09:41.580205 [INF] Listening for client connections on 0.0.0.0:4222
[7048] 2016/12/27 10:09:41.581181 [INF] Server is ready
[7048] 2016/12/27 10:09:41.856760 [INF] STAN: Message store is FILE
[7048] 2016/12/27 10:09:41.856760 [INF] STAN: --------- Store Limits ---------
[7048] 2016/12/27 10:09:41.860668 [INF] STAN: Channels:                 1000
[7048] 2016/12/27 10:09:41.862623 [INF] STAN: -------- channels limits -------
[7048] 2016/12/27 10:09:41.864576 [INF] STAN:   Subscriptions:          1000 *
[7048] 2016/12/27 10:09:41.865554 [INF] STAN:   Messages     :      10000000
[7048] 2016/12/27 10:09:41.867509 [INF] STAN:   Bytes        :     976.56 MB *
[7048] 2016/12/27 10:09:41.869464 [INF] STAN:   Age          :     unlimited *
[7048] 2016/12/27 10:09:41.876305 [INF] STAN: --------------------------------
[7048] 2016/12/27 10:14:12.208363 [INF] 14.161.79.238:56185 - cid:6 - Slow Consumer Detected
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x1 addr=0x10 pc=0x517903]

goroutine 35 [running]:
panic(0x85b900, 0xc04200a090)
        /home/travis/.gimme/versions/go1.7.4.linux.amd64/src/runtime/panic.go:500 +0x1af
github.com/nats-io/nats-streaming-server/stores.(*FileMsgStore).evictFromCache(0xc04201eb60, 0x1493fe6ec25c75d8)
        /home/travis/gopath/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:2174 +0xb3
github.com/nats-io/nats-streaming-server/stores.(*FileMsgStore).backgroundTasks(0xc04201eb60)
        /home/travis/gopath/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:1987 +0x3b1
created by github.com/nats-io/nats-streaming-server/stores.(*FileStore).newFileMsgStore
        /home/travis/gopath/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:1252 +0xca5
```

--- a/stores/filestore.go
+++ b/stores/filestore.go
@@ -2118,6 +2118,10 @@ func (ms *FileMsgStore) addToCache(seq uint64, msg *pb.MsgProto, isNew bool) {
 		c.head = cMsg
 	} else {
 		c.tail.next = cMsg
+		// Ensure last expiration is at least >= previous one.
+		if cMsg.expiration < c.tail.expiration {
+			cMsg.expiration = c.tail.expiration
+		}
 	}
 	cMsg.prev = c.tail
 	c.tail = cMsg
@@ -2135,6 +2139,9 @@ func (ms *FileMsgStore) getFromCache(seq uint64) *pb.MsgProto {
 	if cMsg == nil {
 		return nil
 	}
+	// Bump the expiration
+	cMsg.expiration = time.Now().UnixNano() + cacheTTL
+	// If not already at the tail of the list, move it there
 	if cMsg != c.tail {
 		if cMsg.prev != nil {
 			cMsg.prev.next = cMsg.next
@@ -2146,10 +2153,14 @@ func (ms *FileMsgStore) getFromCache(seq uint64) *pb.MsgProto {
 			c.head = cMsg.next
 		}
 		cMsg.prev = c.tail
+		c.tail.next = cMsg
 		cMsg.next = nil
+		// Ensure last expiration is at least >= previous one.
+		if cMsg.expiration < c.tail.expiration {
+			cMsg.expiration = c.tail.expiration
+		}
 		c.tail = cMsg
 	}
-	cMsg.expiration = time.Now().UnixNano() + cacheTTL
 	return cMsg.msg
 }
 

--- a/stores/filestore_test.go
+++ b/stores/filestore_test.go
@@ -3935,6 +3935,68 @@ func TestBufShrink(t *testing.T) {
 	}
 }
 
+func TestFSCacheList(t *testing.T) {
+	cleanupDatastore(t, defaultDataStore)
+	defer cleanupDatastore(t, defaultDataStore)
+
+	// Increase cacheTTL so eviction does not happen while we test content of list
+	cacheTTL = int64(10 * time.Second)
+	defer func() {
+		cacheTTL = int64(defaultCacheTTL)
+	}()
+
+	fs := createDefaultFileStore(t)
+	defer fs.Close()
+
+	msg := []byte("hello")
+	// Store messages 1, 2, 3
+	for i := 0; i < 3; i++ {
+		storeMsg(t, fs, "foo", msg)
+	}
+
+	cs := fs.LookupChannel("foo")
+	ms := cs.Msgs.(*FileMsgStore)
+
+	// Check list content
+	checkList := func(expectedSeqs ...uint64) {
+		ms.RLock()
+		c := ms.cache
+		cMsg := c.head
+		i := 0
+		good := 0
+		gotStr := ""
+		for cMsg != nil {
+			gotStr = fmt.Sprintf("%v%v ", gotStr, cMsg.msg.Sequence)
+			if cMsg.msg.Sequence == expectedSeqs[i] {
+				good++
+			}
+			i++
+			cMsg = cMsg.next
+		}
+		ms.RUnlock()
+		if i != len(expectedSeqs) || good != len(expectedSeqs) {
+			expectedStr := ""
+			for i := 0; i < len(expectedSeqs); i++ {
+				expectedStr = fmt.Sprintf("%v%v ", expectedStr, expectedSeqs[i])
+			}
+			stackFatalf(t, "Expected sequences: %q, got %q", expectedStr, gotStr)
+		}
+	}
+	// Check that we should have 1, 2, 3
+	checkList(1, 2, 3)
+	// Lookup first, should be moved to end of list
+	ms.Lookup(1)
+	checkList(2, 3, 1)
+	// Repeat...
+	ms.lookup(2)
+	checkList(3, 1, 2)
+	ms.Lookup(3)
+	checkList(1, 2, 3)
+	// Lookup last should leave it there
+	ms.Lookup(3)
+	checkList(1, 2, 3)
+}
+
 func TestFSMsgCache(t *testing.T) {
 	cleanupDatastore(t, defaultDataStore)
 	defer cleanupDatastore(t, defaultDataStore)

